edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent/MetricsFileStorage.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) Microsoft. All rights reserved.
+
+namespace Microsoft.Azure.Devices.Edge.Agent.DiagnosticsComponent
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text;
+    using System.Threading;
+    using Microsoft.Azure.Devices.Edge.Util;
+
+    public class MetricsFileStorage : IMetricsFileStorage
+    {
+        readonly ISystemTime systemTime;
+
+        string directory;
+
+        public MetricsFileStorage(string directory, ISystemTime systemTime = null)
+        {
+            this.directory = directory;
+            this.systemTime = systemTime ?? SystemTime.Instance;
+        }
+
+        public void AddScrapeResult(string data)
+        {
+            Directory.CreateDirectory(this.directory);
+            string file = Path.Combine(this.directory, this.systemTime.UtcNow.Ticks.ToString());
+            File.WriteAllText(file, data);
+        }
+
+        public IDictionary<DateTime, Func<string>> GetData()
+        {
+            return this.GetData(_ => true);
+        }
+
+        public IDictionary<DateTime, Func<string>> GetData(DateTime start)
+        {
+            return this.GetData(ticks => start.Ticks <= ticks);
+        }
+
+        public IDictionary<DateTime, Func<string>> GetData(DateTime start, DateTime end)
+        {
+            return this.GetData(ticks => start.Ticks <= ticks && ticks <= end.Ticks);
+        }
+
+        IDictionary<DateTime, Func<string>> GetData(Func<long, bool> inTimeRange)
+        {
+            return Directory.GetFiles(this.directory)
+                .Select(Path.GetFileName)
+                .SelectWhere(fileName => (long.TryParse(fileName, out long timestamp), timestamp))
```

Comment:
this is very confusing to use datetime as filename.  Should have some format, e.g. edgeMetric_<datetime>.txt, etc.

Reply:
if you use select and then where, linq will delay execution at the end of query; however your current implementation of SelectWhere may not be optimized.

Reply:
I used datetime ticks as the filename so that parsing it would be trivial. 

My SelectWhere is still lazy. Since tryparse has an out param, I would have to parse twice, once for the where and once for the select.

Reply:
alternatively, it could use the query syntax, eg: 
```
long ticks = 0;
            return (IDictionary<DateTime, Func<string>>)from fileName in Directory.GetFiles(this.directory)
                                                        where long.TryParse(Path.GetFileName(fileName), out ticks) && inTimeRange(ticks)
                                                        select new KeyValuePair<DateTime, Func<string>>(new DateTime(ticks), this.GetFileFunc(ticks));
```